### PR TITLE
[SPARK-50331][INFRA] Add a daily test for PySpark on MacOS

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims to add a daily test for PySpark on macos-15

### Why are the changes needed?
Check PySpark on macos-15. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: .github/workflows/python_macos_test.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      - name: Install Java ${{ matrix.java }}
+        uses: actions/setup-java@v4
+        with:
+          distribution: zulu
+          java-version: ${{ matrix.java }}
+      - name: Install Python packages (Python ${{matrix.python}})
+        run: |
+          python${{matrix.python}} -m pip install --ignore-installed 'blinker>=1.6.2'
+          python${{matrix.python}} -m pip install --ignore-installed 'six==1.16.0'
+          python${{matrix.python}} -m pip install py-cpuinfo && \
```
Comment (LuciferYang) [inline .github/workflows/python_macos_test.yml:136]: Added `pip install py-cpuinfo` to fix

```
Collecting deepspeed
  Downloading deepspeed-0.15.4.tar.gz (1.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 50.9 MB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [22 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/setup.py", line 41, in <module>
          from op_builder.all_ops import ALL_OPS, accelerator_name
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/op_builder/all_ops.py", line 29, in <module>
          builder = get_accelerator().create_op_builder(member_name)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/accelerator/mps_accelerator.py", line 244, in create_op_builder
          builder_class = self.get_op_builder(op_name)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/accelerator/mps_accelerator.py", line 251, in get_op_builder
          from deepspeed.ops.op_builder.cpu import NotImplementedBuilder
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/deepspeed/__init__.py", line 25, in <module>
          from . import ops
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/deepspeed/ops/__init__.py", line 6, in <module>
          from . import adam
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/deepspeed/ops/adam/__init__.py", line 6, in <module>
          from .cpu_adam import DeepSpeedCPUAdam
        File "/private/var/folders/7x/5h04sv9566g7qmpwbf47vwb00000gn/T/pip-install-qryq0e3d/deepspeed_96db0a89d65b4f5a8785ab501864592b/deepspeed/ops/adam/cpu_adam.py", line 7, in <module>
          from cpuinfo import get_cpu_info
      ModuleNotFoundError: No module named 'cpuinfo'
      [end of output]
```

**Comment by LuciferYang** [comment]
- https://github.com/LuciferYang/spark/actions/runs/11854044932/job/33035384447

The test was largely successful, except for the modules related to mllib, which seems to need some fix to the test cases on macos. I think we can take care of this in the followup.

```
 test_data_loader (pyspark.ml.tests.connect.test_parity_torch_data_loader.TorchDistributorBaselineUnitTestsOnConnect.test_data_loader) ... WARNING: Using incubator modules: jdk.incubator.vector
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Started distributed training with 1 executor processes
W1115 13:42:15.675000 65006 torch/distributed/elastic/multiprocessing/redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
W1115 13:42:15.675000 65006 torch/distributed/elastic/multiprocessing/redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
Traceback (most recent call last):
  File "/Users/runner/work/spark/spark/python/target/5c60ac90-7dd0-42ad-9c5f-4700f5cf9926/tmpg209ldzx/train.py", line 8, in <module>
    output = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/spark/spark/python/pyspark/ml/torch/tests/test_data_loader.py", line 71, in train_function
    return list(data_loader)
           ^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/torch/utils/data/dataloader.py", line 484, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/torch/utils/data/dataloader.py", line 415, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/torch/utils/data/dataloader.py", line 1138, in __init__
    w.start()
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/opt/homebrew/Cellar/python@3.11/3.11.10/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object '_SparkPartitionTorchDataset._get_field_converter.<locals>.converter'
E1115 13:44:37.049000 65006 torch/distributed/elastic/multiprocessing/api.py:869] failed (exitcode: 1) local_rank: 0 (pid: 66724) of binary: /opt/homebrew/opt/python@3.11/bin/python3.11
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/opt/homebrew/lib/python3.11/site-packages/torch/distributed/run.py", line 923, in <module>
    main()
  File "/opt/homebrew/lib/python3.11/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 355, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/torch/distributed/run.py", line 919, in main
    run(args)
  File "/opt/homebrew/lib/python3.11/site-packages/torch/distributed/run.py", line 910, in run
    elastic_launch(
  File "/opt/homebrew/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 269, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError:
============================================================
/Users/runner/work/spark/spark/python/target/5c60ac90-7dd0-42ad-9c5f-4700f5cf9926/tmpg209ldzx/train.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-11-15_13:44:37
  host      : Mac-1731662073710.local
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 66724)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
```

The other few failures were due to duplicate naming of 'Upload test results to report'.

**Comment by LuciferYang** [comment]
cc @HyukjinKwon @dongjoon-hyun @zhengruifeng @panbingkun FYI

File: .github/workflows/python_macos_test.yml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+# specific language governing permissions and limitations
+# under the License.
+#
+
+name: Build and test using Maven
+
+on:
+  workflow_call:
+    inputs:
+      java:
```
Comment (zhengruifeng) [inline .github/workflows/python_macos_test.yml:25]: it is a bit more complicated than I thought, it seems we simply specify the java/hadoop version / branches in other python daily jobs.

https://github.com/apache/spark/blob/9d79ab42b127d1a12164cec260bfbd69f6da8b74/.github/workflows/build_python_3.10.yml#L34

Comment (zhengruifeng) [reply .github/workflows/python_macos_test.yml:25]: or should we specify the versions in another file?

**Review by zhengruifeng — COMMENTED** [review]
In General LGTM, thank you so much

File: .github/workflows/python_macos_test.yml (hunk 2 above)
Comment (LuciferYang) [reply .github/workflows/python_macos_test.yml:25]: 1. Only the content defined here can specify versions in other scripts. We can see that `build_and_test.yml` also defines this content.
2. Actually, for content that is the same as the default version, it is not necessary to specify it again in the sub-script."So it's okay not to specify 

```
     java: 17
      branch: master
      hadoop: hadoop3
``` 

again in `build_python_3.10.yml`

**Review by zhengruifeng — APPROVED** [review]

**Comment by panbingkun** [comment]
LGTM

**Comment by LuciferYang** [comment]
image

The failed case is unrelated to the current PR, and @yaooqinn  is working on fixing the issue. I will merge this PR and observe the test results. If there are any other issues, I can address them in the followup.

**Comment by LuciferYang** [comment]
Merged into master. Thanks @zhengruifeng and @panbingkun

**Pull request closed without merging**